Update to add SCID-less HTTP URLs, control over portability, and combining versionId-entryHash (#72)
bcgov/trustdidweb: spec/specification.md

the prefix, is the [method-specific identifier](#method-specific-identifier),
specified below.

> ### Method Specific Identifier

Reviewer: ```suggestion
### Method-Specific Identifier
```

Reviewer: That is the term in DID Core.. also some references below

Author: Sounds good.

Author: Fixed throughout.

### Method-specific Identifier

The method specific identifier contains both the [[ref: self-certifying
identifier]] (SCID) for the DID, and a fully qualified domain name that is
secured by a TLS/SSL certificate, with an optional path to the [[ref: DID Log]].
As per the Augmented Backus-Naur Form (ABNF) notation below, the SCID **MUST**
be a part of the method specific identifier by one of the following mechanisms:

- A separate, colon-delimited element at the start of the method specific identifier.
- Embedded in the fully qualified domain name.
- Embedded in the optional path relative to the fully qualified domain name.
  
Formal rules describing valid domain name syntax are described in
[[spec:RFC1035]], [[spec:RFC1123]], and [[spec:RFC2181]]. Each `did:tdw` DID's
globally unique [[ref: SCID]] is [generated](#scid-generation-and-verification)

> during the creation the DID based on the DID's initial content and placed into

Reviewer 2: ```suggestion
during the creation of the DID based on it's initial content and placed into
```

The location of the `did:tdw` `did.jsonl` [[ref:DID Log]] file is the same as
where the comparable `did:web` `did.json` file is published. A [[ref: DID
Controller]] **MAY** choose to publish both DIDs and so, both files. The process
to do so is described in the [publishing a parallel `did:web`
DID](#publishing-a-parallel-didweb-did) section of this specification.

### The DID Log File

The DID log file contains a list of entries, one for each version of the DID. A
version of the DID is an update to the contents of the resolved DIDDoc for the
DID, or a change to the [[ref: parameters]] that control the generation and
verification of the DID.

Each entry is a JSON array consisting of the following 5 items.

> `[ versionId, versionTime, parameters, DIDDoc, Data Integrity Proof ]`

Reviewer: ```suggestion
`[ versionId, versionTime, parameters, DIDDoc, proofs ]`
```
proofs being an array

Reviewer: Maybe we could rename 'DIDDoc' to 'state' here to avoid confusion?

Author: Hmmm...agree maybe "DIDDoc State"?  I like that it references DIDDoc, and think just state is a bit ambiguous, but just DIDDoc is not correct.  I'll change it to "DIDDoc State" -- but feel free to push harder.

Author: Fixed.

      key pair's public key. The [[ref: multikey]] representations of the public
      keys are placed in the `updateKeys` item in [[ref: parameters]].
   3. The public key(s) of the authorization key pair(s) **MAY** be used in the
      DIDDoc as well, but that is not required.
3. Create the initial DIDDoc for the DID, with whatever content is required. The
   top level `id` item **MUST** be included and **MUST** be the DID string from
   step 1, including the placeholder for the SCID. All other absolute
   self-reference's to the DID in the DIDDoc must use the form defined in step 1, with
   the identified placeholder for the [[ref: SCID]]  (e.g.,
   `did:tdw:example.com:{SCID}#key-1`, `did:tdw:{SCID}:example.com#key-1`, etc.).
4. Define the JSON object using valid [[ref: parameters]] to configure the DID generation
   and verification processes.
   The [DID Generation and Verification Parameters](#didtdw-did-method-parameters)
   section of this specification defines the permitted [[ref: parameters]]. That
   section defines what items **MUST** be included in this first log entry for the DID.
5. Formulate an input JSON array containing the following items for processing:

>    `[ "1-{SCID}", "<current time>", "parameters": [ <parameters>], { "value": "<DIDDoc with Placeholders>" } ]`

Reviewer: I believe we would stick with "{SCID}" as the first (non-published) entry hash. After hashing it would be replaced with "1-{entryHash}"

Author: Obvs it is a toss up and has no impact.  I'll change it back -- it will probably be a little easier to describe.

Author: Fixed.

   pre-publication JSON array of what will become the first log entry. The
   placeholder is the literal string "`{SCID}`".

>    - The `versionId` entry, which must be `1-{SCID}`, which is the first
>      version number for the DID `1`, a dash `-`, and the placeholder for the
>      SCID.

Reviewer: ```suggestion
   - The `versionId` entry, which must be `{SCID}`, the placeholder for the SCID.
```

2. Determine the length of the `scid` and save it for use later. If the length
   is less than 28 characters, terminate the resolution process with an error.
3. Remove from the first [[ref: DID log entry]] the data integrity proof.

> 4. Replace the `versionId` value with the literal `"1-{SCID}"`.

Reviewer: ```suggestion
4. Replace the `versionId` value with the literal `"{SCID}"`.
```

Author: Doh…I thought I had done a search. Fixed.

To verify the `entryHash` for a given  `did:tdw` [[ref: DID log entry]], a DID
Resolver **MUST** execute the following process:

1. Extract the `versionId` (first item) in the [[ref: DID log entry]], and
   remove from it the version number and dash prefix, leaving the log entry
   `entryHash`.
2. Remove the [[ref: Data Integrity]] proof (6th item) from the log entry.
3. Set the first item of the entry to the `versionId` (first item) of the
   previous log entry. If this is the first entry in the log, set the value to

>    the `1-<scid>` where `<scid>` if the SCID of the DID.

Reviewer: ```suggestion
   the [[ref: SCID]] for the DID.
```

Author: Fixed.
